tests/rptest/services/action_injector.py

```diff
@@ -0,0 +1,330 @@
+
+
+class DisruptiveAction:
+    """
+    Defines an action taken on a node or on the cluster as a whole which causes a disruption.
+
+    The action could be a process failure, leadership transfer, topic modification etc.
+
+    The action can be reversible, it also stores the set of affected nodes and the last node
+    the action was applied on.
+    """
+    def __init__(self, redpanda: RedpandaService, config: ActionConfig,
+                 admin: Admin):
+        self.admin = admin
+        self.config = config
+        self.redpanda = redpanda
+        self.affected_nodes = set()
+        self.is_reversible = False
+        self.last_affected_node = None
+
+    def max_affected_nodes_reached(self) -> bool:
+        """
+        Checks if the number of affected nodes so far equals the maximum number of nodes
+        this action is allowed to affect. If so all future calls to action will be no-op.
+        """
+        raise NotImplementedError
+
+    def target_node(self) -> ClusterNode:
+        """
+        Randomly selects the next node to apply the action on. A set of affected
+        nodes is maintained so that we do not apply the action on nodes which were
+        already targeted in previous invocations.
+        """
+        available = set(self.redpanda.nodes) - self.affected_nodes
+        if available:
+            selected = random.choice(list(available))
+            names = {n.account.hostname for n in available}
+            self.redpanda.logger.info(
+                f'selected {selected.account.hostname} of {names} for operation'
+            )
+            return selected
+
+    def do_action(self) -> ClusterNode:
+        """
+        Applies the disruptive action, returns node or entity the action was applied on
+        """
+        raise NotImplementedError
+
+    def action(self) -> ClusterNode:
+        if not self.max_affected_nodes_reached():
+            return self.do_action()
+
+    def do_reverse_action(self) -> ClusterNode:
+        """
+        Reverses the last applied action if applicable.
+        """
+        raise NotImplementedError
+
+    def reverse(self) -> ClusterNode:
+        if self.is_reversible and self.last_affected_node is not None:
+            return self.do_reverse_action()
+
+    def do_restore_nodes(self, nodes_to_restore: Set[ClusterNode]):
+        raise NotImplementedError
+
+    def restore_state_on_exit(self, action_log: List[ActionLogEntry]):
+        """
+        Optionally restore state when the action injector thread is ending.
+        Uses the action log to determine what restoration should be done.
+        """
+        node_final_state = defaultdict(lambda: False)
+        for entry in action_log:
+            node_final_state[entry.node] = entry.is_reverse_action
+
+        nodes_to_restore = {
+            node
+            for node, is_reversed in node_final_state.items()
+            if not is_reversed
+        }
```

Comment:
The map is to capture the final state of each node and to avoid restore on nodes which do not need it. If action log looks like this

```
[
 (NodeA, Down),
 (NodeA, Up),
 (NodeB, Down),
 (NodeB, Up),
 (NodeC, Down),
]
```

then the map will look like `{NodeA: Up, NodeB: Up, NodeC: Down}` and the set will contain just NodeC, which is the only node to bring up.

It could also be possible to just look at the last entry in action log, but it is possible that some recovery has failed at some point and multiple nodes are down, so it is safer to look at the entire list, calculate which actions were finally not reversed, and reverse them.